### DISCUSSION TITLE
$transaction failure with TypeError: Cannot assign to read only property `_transactionId`

### DISCUSSION THREAD

**Discussion by jeffdc** [prisma/prisma#6395] [Q&A] [answered]
I have spent way longer than I care to admit (or even think about) trying to figure out what is going on with this.

I have some simple `$transaction()` calls that worked fine in 2.13.1. As soon as I update to a version above 2.13 (I have tried 2.14 and then 2.20) they fail with an error message:

```
/Users/jeff/dev/gallformers/node_modules/@prisma/client/runtime/index.js:34366
      return this._transactionId++;
                                ^
TypeError: Cannot assign to read only property '_transactionId' of object '#<NewPrismaClient>'
    at NewPrismaClient.getTransactionId (/Users/jeff/dev/gallformers/node_modules/@prisma/client/runtime/index.js:34366:33)
    at NewPrismaClient.$transactionInternal (/Users/jeff/dev/gallformers/node_modules/@prisma/client/runtime/index.js:34377:34)
    at NewPrismaClient.$transaction (/Users/jeff/dev/gallformers/node_modules/@prisma/client/runtime/index.js:34397:21)
...
```

I then rolled back to the old version of Prisma that I was using. 2.13.1 and it works fine again.

Here is an example of a simple call that I constructed for testing purposes that fails:
``` Typescript
export const testTx = (): Promise<[species[]]> => {
    return db.$transaction([db.species.findMany({ where: { name: { contains: 'alba' } } })]);
};
```
Things I have done:
- I have made sure that my client has been regenerated
- Checked TS version, 4.2.3
- Searched the internet for similar issues
- Looked at the [source code](https://github.com/prisma/prisma/blob/974cbeff4a7f616137ce540d0ec88a2a86365892/src/packages/client/src/runtime/getPrismaClient.ts#L882) in question and it seems fine. A `private` `class` member that is *not* declared `readonly` should be able to be incremented with `++` from within the class and JS/TS operator precedence means that the class member is "dereferenced" before being incremented**
- Deleted all generated artifacts etc. (e.g., node_modules, .next) and re-built (went so far as to reclone in a diff directory)
- Tested other Prisma functionality and it seems to be working without issue

I do not see anyone else having this issue (searched high and low). So it must be something dumb on my side. I have isolated it down to the simplest case and it still fails, every time.

** As an aside I cringed at a bit at the pattern of using the method name `getTransactionId`, which looks the world like a "getter" but the code actually increments the internal `_transactionId` before returning the state value, which breaks the model that a getter would be idempotent.

**Comment by jeffdc** [comment]
I am still at a loss. I think that some combination of the property being `private` and there being a getter that mutates the state is at the core of the issue. I hacked the generated prisma client to change `_transactionId` from a `number` to a `number[]` so that the mutation of the number within the array (rather than the property itself) would bypass this error. This has at least let me start testing the original reason that I was trying to update Prisma from 2.13.1 to a newer version (transaction ordering). 

Everything else I have: my code, 3rd party libs, etc. works. Prisma version > 2.13.1 however does not.

@timsuchanek It looks like you authored this code in the past several months. Do you have any ideas?

**Reply by janpio** [reply]
Can't help with the error, but to reproduce this I would create a new project with just this model, and then try to execute this query in an index.ts file directly. If this does not trigger the same error, something along the way has to go wrong.

**Comment by jeffdc** [comment, marked as answer]
I finally got to the bottom of this... 

Months ago in trying to deal with ramifications of #5007 I added code to put my DB connection (in dev only) into `globalThis` and then `Object.freeze()` it. When the code was committed to add transaction ordering (#4759) things broke. 

As I mentioned above in my original write-up.I believe that the code in #4579 is flawed. The problem is that the [getter](https://github.com/prisma/prisma/blob/974cbeff4a7f616137ce540d0ec88a2a86365892/src/packages/client/src/runtime/getPrismaClient.ts#L882) on `_transactionId` is not idempotent, it modifies the internal state on every call. When I called `Object.freeze(db)` this made the `_transactionId` read-only and then boom.

I realize that this problem only exists when you `freeze` the object and the fix is to just not do that 😄 , but mutating state in the getter is not good practice and it might be worth looking at changing that.

If nothing else, I hope that this write-up helps prevent someone else suffering this pain in the future!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
